pandas/core/indexes/interval.py

```diff
@@ -1101,9 +1101,25 @@ def interval_range(
     breaks: np.ndarray | TimedeltaIndex | DatetimeIndex
 
     if is_number(endpoint):
+        dtype: Any = np.dtype("int64")
         if com.all_not_none(start, end, freq):
+            if (
+                isinstance(start, (float, np.float16))
+                or isinstance(end, (float, np.float16))
+                or isinstance(freq, (float, np.float16))
+            ):
+                dtype = np.dtype("float64")
+            elif isinstance(start, int) or isinstance(end, int):
+                dtype = np.dtype("int64")
+            elif np.issubdtype(start.dtype, np.integer) and np.issubdtype(
```

Comment:
Can simplify a lot of these cases by checking

```
if isinstance(start, (np.integer, np.floating)) and isinstance(end, (np.integer, np.floating)) and start.dtype == end.dtype:
    dtype = start.dtype
```

Reply:
Done!